src/realm/sync/changeset.hpp

```diff
@@ -539,7 +539,13 @@ inline StringData Changeset::string_data() const noexcept
 
 inline StringBufferRange Changeset::append_string(StringData string)
 {
-    m_string_buffer->reserve(1024); // we expect more strings
+    // small string optimization; we expect more strings, but constantly requesting many small allocation increases
+    // to the string buffer can become a performance bottleneck in highly fragmented or low memory devices
+    constexpr size_t small_string_buffer_size = 1024;
+    if (m_string_buffer->capacity() - m_string_buffer->size() < string.size() &&
+        string.size() < small_string_buffer_size) {
+        m_string_buffer->reserve(m_string_buffer->capacity() + small_string_buffer_size);
+    }
```

Comment:
I was going to suggest this exact change, but then I thought "wait don't vectors/strings amortize their growth?" [At least on libc++ strings double in size when they exceed their capacity until you reach a capacity that's half the size of the maximum capacity of the string](https://github.com/llvm/llvm-project/blob/6fa9120080c35a5ff851c3fc3358692c4ef7ce0d/libcxx/include/string#L2330). Then I read the cppreference page for reserve() and realized we're probably shooting ourselves in the foot here. https://en.cppreference.com/w/cpp/string/basic_string/reserve. Until c++20:
* If new_cap is less than the current capacity(), this is a non-binding shrink request.
* If new_cap is less than the current [size()](https://en.cppreference.com/w/cpp/string/basic_string/size), this is a non-binding shrink-to-fit request equivalent to [shrink_to_fit()](https://en.cppreference.com/w/cpp/string/basic_string/shrink_to_fit) (since C++11).

I suspect that what's going on here is that we reserve 1024, and then for every additional string that we append past the first 1024 bytes, we [effectively shrink_to_fit the string](https://github.com/llvm/llvm-project/blame/6d020a5ac2d5e31ab8d472e139e9caec405a5006/libcxx/include/string#L3268). If we change this to only call reserve if the capacity is less than 1024, do you still see the slowdown?

Btw, this behavior goes away in C++20.

Reply:
Huh, I wouldn't have guessed that `reserve()` would do a `shrink_to_fit()`. I guess I'm glad c++20 fixes that.

Reply:
Interesting, good analysis. I will change this to let the string implementation do its own capacity management and ask Eric to see if it is more effective.